lib/Basics/Exceptions.h

```diff
@@ -35,43 +35,89 @@
-  throw arangodb::basics::Exception(code, message, __FILE__, __LINE__)
+  throw ::arangodb::basics::Exception(code, message, __FILE__, __LINE__)
 
 /// @brief throws an arango result if the result fails
-#define THROW_ARANGO_EXCEPTION_IF_FAIL(expression)                                        \
-  do {                                                                                    \
-    auto&& expressionResult = (expression);                                               \
-    if (expressionResult.fail()) {                                                        \
-      throw arangodb::basics::Exception(std::move(expressionResult), __FILE__, __LINE__); \
-    }                                                                                     \
+#define THROW_ARANGO_EXCEPTION_IF_FAIL(expression)                                          \
+  do {                                                                                      \
+    auto&& expressionResult = (expression);                                                 \
+    if (expressionResult.fail()) {                                                          \
+      throw ::arangodb::basics::Exception(std::move(expressionResult), __FILE__, __LINE__); \
+    }                                                                                       \
   } while (0)
 
-namespace arangodb {
-namespace basics {
+// Fix MSVC's preprocessor...
+#define EXPAND_(args) args
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+
+#define THROW_AND_CATCH_(THROW_MACRO, ...)                                  \
+  do {                                                                      \
+    try {                                                                   \
+      EXPAND_(THROW_MACRO(__VA_ARGS__));                                    \
+    } catch (::arangodb::basics::Exception const& ex) {                     \
+      LOG_TOPIC("fa7a1", FATAL, ::arangodb::Logger::CRASH) << ex.message(); \
+      TRI_ASSERT(false);                                                    \
+    } catch (...) {                                                         \
+      TRI_ASSERT(false);                                                    \
+    }                                                                       \
+  } while (0)
+
+#define ASSERT_OR_THROW_ARANGO_EXCEPTION(...) \
+  THROW_AND_CATCH_(THROW_ARANGO_EXCEPTION, __VA_ARGS__)
+
+#define ASSERT_OR_THROW_ARANGO_EXCEPTION_PARAMS(...) \
+  THROW_AND_CATCH_(THROW_ARANGO_EXCEPTION_PARAMS, __VA_ARGS__)
+
+#define ASSERT_OR_THROW_ARANGO_EXCEPTION_FORMAT(...) \
+  THROW_AND_CATCH_(THROW_ARANGO_EXCEPTION_FORMAT, __VA_ARGS__)
+
+#define ASSERT_OR_THROW_ARANGO_EXCEPTION_MESSAGE(...) \
+  THROW_AND_CATCH_(THROW_ARANGO_EXCEPTION_MESSAGE, __VA_ARGS__)
+
+#define ASSERT_OR_THROW_ARANGO_EXCEPTION_IF_FAIL(...) \
+  THROW_AND_CATCH_(THROW_ARANGO_EXCEPTION_IF_FAIL, __VA_ARGS__)
+
+#else
+
+#define ASSERT_OR_THROW_ARANGO_EXCEPTION(...) \
```

Comment:
I am not really a big fan of these macros.
Do I understand correctly that in maintainer mode, they will throw an exception and terminate the program (but not in case no exception occurs...?).
And in release mode they will simply throw, but not end the program?
The macro names indicate that it's going to be either an assertion or a throw here, but I am wondering what the use case is for these macros? IMO we shouldn't add more if not absolutely required. Maybe there is a use case for such macros, but I don't see it yet.

Reply:
It's something I do often: Add an assertion in maintainer mode, and throw an exception in release builds, e.g
```
if (<condition>) {
  TRI_ASSERT(false);
  THROW_ARANGO_EXCEPTION_MESSAGE(<code>, <message>);
}
```
. Essentially, in many, if not most, places where I add an assertion (in maintainer mode), I also want to throw an exception in release mode. This also makes sure the specific error is logged in maintainer mode as well, which is lost in the example above.

They are not absolutely required, though I do think I would probably use them quite often now if available.